ouroboros-consensus/docs/test-consensus/Plan.md

```diff
@@ -0,0 +1,629 @@
+immediately after forging a new block and never otherwise. Each message carries
+the new chain to every other node (actually to the sender itself too, but that
+seems ultimately inconsequential). There are no other types of messages.
+Messages essentially arrive _between_ slots, since if they arrive during a
+slot, they are ignored until the onset of the next slot (see discussion and
+pseudo-code in the The π_SPoS Protocol section above).
+
+Many definitions, theorems, and proofs in the paper refer to the maximum
+message delay Δ. If Δ is 0, messages arrive at the end of the same slot they
+are sent; if Δ is 1, they arrive no later than the end of the subsequent slot;
+and so on. Different messages carrying the same chain may arrive at different
+parties after different delays, but no delay will exceed Δ. The parties and
+nodes themselves are both unaware of Δ, so they aren't, for example, using it
+to check if a sent message has "timed-out". The theorems and proofs, on the
+other hand, do explicitly rely on Δ.
+
+The implementation's messages are very different than the paper's for practical
+reasons that are lower-level than the paper addressed, such as keeping the
+message size relatively small. Asymmetric client-server pairs of nodes
+communicate via _mini protocols_, most notably ChainSync and BlockFetch. The
+two nodes in a mini protocol instance exchange messages of various types, each
+of which corresponds to an action in a labeled transition system that specifies
+the mini protocol. These messages are all much smaller than a whole chain; the
+largest message contains exactly one block. These mini protocols cumulatively
+achieve the same effect as a message from the paper: communication of a forged
+chain to the other nodes in the network. However, the mini protocols do it via
+a sequence of small, generic handshakes between two nodes and along paths of
+node-to-node edges in a connected forwarding network that may have fewer edges
+than the paper's fully connected topology. As a concrete contrast, the paper's
+parties only ever send a block immediately after they forge it, whereas the
+nodes in a _non-fully_ connected network must relay blocks they did not
+themselves forge.
+
+The delay of a single mini protocol message is only one part of the whole sum
+that is analogous to the delay of a paper message. The test suite therefore
+considers a _message_ to be a pair of a block and a node and such a message's
+_delay_ to be the difference between the slot in which the block was forged and
+the slot in which it was added to the node's chain database. This definition
+preserves the property from the paper that a message delay is the duration (in
+slots) of the communication of a forged chain (headed by the new block) to
+another node in the network.
+
+(Remark. It is possible for different nodes to forge the same EBB in either the
+same or different slots. This particular definition of message delay therefore
```

Comment:
I thought an EBB for a certain epoch only fits into a certain slot, i.e. the first slot of the next epoch (this slot can *also* contain a regular block). I might be mistaken, but I was under the impression that each (core) node produces every EBB "automically" at the end of each epoch.